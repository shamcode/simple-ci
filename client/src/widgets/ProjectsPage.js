import { options } from 'sham-ui';
import ProjectsPageTemplate from './ProjectsPage.sht';

export default class ProjectsPage extends ProjectsPageTemplate {
    @options get dataLoaded() { return false;  }
    @options get projects() { return []; }
    @options get errors() { return []; }

    @options
    afterRender() {
        this._loadPageData();
    }

    _loadPageData() {
        const loading = new Promise( ( resolve ) => {
            setTimeout( () => resolve( {
                projects: [
                    { name: 'Fixture' },
                    { name: 'Second fixture' },
                    { name: 'Simple CI' },
                    { name: 'Simple CI 1' },
                    { name: 'Simple CI 2' },
                    { name: 'Snackbars are often used as tooltips/popups' }
                ]
            } ), 500 );
        } );
        loading.then(
            ::this._loadedPageDataSuccess,
            ::this._loadedPageDataFail
        ).finally(
            () => this.update( { dataLoaded: true } )
        )
    }

    _loadedPageDataSuccess( { projects } ) {
        this.update( {
            projects,
            errors: []
        } )
    }

    _loadedPageDataFail() {
        this.update( {
            projects: [],
            errors: [ "Fail loaded project!" ]
        } )
    }
}