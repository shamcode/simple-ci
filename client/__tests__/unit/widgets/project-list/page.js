import { DI } from 'sham-ui';
import ProjectsPage from '../../../../src/widgets/project-list/page';
import * as directives from 'sham-ui-directives';
import renderer from 'sham-ui-test-helpers';

afterEach( () => {
    DI.bind( 'store', null );
    DI.bind( 'router', null );
} );

it( 'renders correctly', () => {
    const storeMock = jest.fn().mockReturnValue( Promise.resolve( { projects: [] } ) );
    DI.bind( 'store', {
        getProjects: storeMock
    } );
    DI.bind( 'router', {
        generate: jest.fn().mockReturnValueOnce( '/' )
    } );

    const meta = renderer( ProjectsPage, { directives } );
    expect( storeMock.mock.calls ).toHaveLength( 1 );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'render errors', () => {
    const storeMock = jest.fn().mockReturnValue( Promise.reject() );
    DI.bind( 'store', {
        getProjects: storeMock
    } );
    DI.bind( 'router', {
        generate: jest.fn().mockReturnValueOnce( '/' )
    } );

    const meta = renderer( ProjectsPage, { directives } );
    expect( meta.toJSON() ).toMatchSnapshot();
} );
