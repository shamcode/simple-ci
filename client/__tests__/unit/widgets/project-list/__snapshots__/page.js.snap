// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render errors 1`] = `
Object {
  "ID": "widget",
  "constructorOptions": Object {
    "directives": Object {
      "EventListener": [Function],
      "disabled": [Function],
      "onclick": [Function],
      "oninput": [Function],
      "onload": [Function],
      "onresize": [Function],
      "onscroll": [Function],
      "onselect": [Function],
      "onsubmit": [Function],
    },
  },
  "html": "
<!--if-->
<!--FlushPageDataLoading-->
<!--if-->
",
  "nodes": Array [
    Object {
      "ID": undefined,
      "constructorOptions": null,
      "html": null,
      "nodes": null,
      "options": null,
    },
  ],
  "options": Object {
    "actionSequence": Array [
      "render",
      "bindEvents",
    ],
    "afterRender": [Function],
    "dataLoaded": false,
    "directives": Object {
      "EventListener": [Function],
      "disabled": [Function],
      "onclick": [Function],
      "oninput": [Function],
      "onload": [Function],
      "onresize": [Function],
      "onscroll": [Function],
      "onselect": [Function],
      "onsubmit": [Function],
    },
    "errors": Array [],
    "projects": Array [],
    "renderAsync": false,
    "renderDependence": Array [],
    "types": Array [],
  },
}
`;

exports[`renders correctly 1`] = `
Object {
  "ID": "widget",
  "constructorOptions": Object {
    "directives": Object {
      "EventListener": [Function],
      "disabled": [Function],
      "onclick": [Function],
      "oninput": [Function],
      "onload": [Function],
      "onresize": [Function],
      "onscroll": [Function],
      "onselect": [Function],
      "onsubmit": [Function],
    },
  },
  "html": "
<!--if-->
<!--FlushPageDataLoading-->
<!--if-->
",
  "nodes": Array [
    Object {
      "ID": undefined,
      "constructorOptions": null,
      "html": null,
      "nodes": null,
      "options": null,
    },
  ],
  "options": Object {
    "actionSequence": Array [
      "render",
      "bindEvents",
    ],
    "afterRender": [Function],
    "dataLoaded": false,
    "directives": Object {
      "EventListener": [Function],
      "disabled": [Function],
      "onclick": [Function],
      "oninput": [Function],
      "onload": [Function],
      "onresize": [Function],
      "onscroll": [Function],
      "onselect": [Function],
      "onsubmit": [Function],
    },
    "errors": Array [],
    "projects": Array [],
    "renderAsync": false,
    "renderDependence": Array [],
    "types": Array [],
  },
}
`;
